Add stats log reader

bogdan-szabo-sociomantic opened the pull request.

bogdan-szabo-sociomantic committed:
Add stats log reader

nemanja-boric-sociomantic commented on src/ocean/util/log/Stats.d:
```diff
@@ -650,3 +656,369 @@ unittest
+
+    static public StatsLine opCall (cstring line)
+    {
+        static Buffer!(char) buffer_line;
```
Do you need this static variable? It feels like you could do all this just by slicing the `line`?

nemanja-boric-sociomantic commented on src/ocean/util/log/Stats.d:
```diff
@@ -650,3 +656,369 @@ unittest
+
+**************************************************************************/
+
+public struct StatsLine
```
`StatsLine` could just hold line as `cstring` and just slice the `line` passed in `opCall`. That would mean that the lifetime of `StatsLine` is valid as long as `line` passed to `opCall` is valid, but I think that is fine. AFAICS, you don't need to pass the `StatsLine` around, and if you want you could provide `dup` method which would return you a copy of `StatsLine` which would manage its own memory.

That way I think you've removed all allocations here.

nemanja-boric-sociomantic commented on src/ocean/util/log/Stats.d:
```diff
@@ -650,3 +656,369 @@ unittest
+
+        stats_line.line = buffer_line[time_end_position .. len];
+
+        return stats_line;
```
So, in this method you always create new `StatsLine` and pass it to the caller. Creating `StatsLine` involves creating `Buffer` which `StatsLine` doesn't reuse in any way, and since it's private nobody else can do reuse it, so having `Buffer` there makes a little point (except to prevent copying StatsLines).

I think we should you could/should do it in a way where `StatsLine` is just a _view_ on the line.

bogdan-szabo-sociomantic commented:
@nemanja-boric-sociomantic ready for another look

mihails-strasuns commented on src/ocean/util/log/Stats.d:
```diff
@@ -650,3 +656,402 @@ unittest
+    }
+    catch(Exception e)
+    {
+        test!("==")(e.message, "The stats are empty");
```
Why not simply `testThrown`?

bogdan-szabo-sociomantic replied:
because `testThrown` can not check the error message. I updated the code to use `testThrown` anyway... It looks like there are no tests that check for the error messages

mihails-strasuns replied:
Yeah we don't normally care about exact error message and it is not worth extra lines of boilerplate. If specific exception kind is crucial, it should be denoted by creating dedicated exception class and doing `testThrown!(MyExceptionClass)` instead.

mihails-strasuns commented on src/ocean/util/log/Stats.d:
```diff
@@ -650,3 +656,402 @@ unittest
+    catch(Exception e)
+    {
+        test!("==")(e.message, "missing_key is not present in the stats");
+        has_exception = true;
```
ditto `testThrown`

mihails-strasuns reviewed (approved):
LGTM apart from few nitpicks. One proposal to consider is to move it new `ocean.util.log.StatsReader` module, I personally like small independent modules more.

bogdan-szabo-sociomantic commented:
@mihails-strasuns @nemanja-boric-sociomantic  ready for another look. I moved the code to a new module and refactored the tests to use the testThrown function

nemanja-boric-sociomantic commented on src/ocean/util/log/StatsReader.d:
```diff
@@ -0,0 +1,393 @@
+    Refer to the class' description for information about their actual usage.
+
+    Copyright:
+        Copyright (c) 2009-2018 Sociomantic Labs GmbH.
```
```
        Copyright (C) 2018 dunnhumby Germany GmbH. All rights reserved
```

nemanja-boric-sociomantic commented on src/ocean/util/log/StatsReader.d:
```diff
@@ -0,0 +1,393 @@
+
+    Classes that can be used to parse the graphite stats.
+
+    Currently, `StatsLogReader` parses an `InputStream` to `StatsLine`. The lines
```
No need for `Currently,` here, just say `Parses an InputStream...`

nemanja-boric-sociomantic commented:
Thanks!
